inst/shiny/modules/tab_visuals.R

```diff
@@ -0,0 +1,560 @@
+      param_choices <- data() %>%
+        pull(ANALYTE) %>%
+        unique()
+
+      pickerInput(
+        ns("generalplot_analyte"),
+        "Select Analyte:",
+        choices = param_choices,
+        selected = param_choices[1],
+        multiple = FALSE,
+        options = list(`actions-box` = TRUE)
+      )
+
+    })
+
+    # select the usubjid for the general lineplot
+    output$generalplot_usubjid <- renderUI({
+      # deselect choices that are no pp parameters
+      param_choices <- data() %>%
+        pull(USUBJID) %>%
+        unique()
+
+      pickerInput(
+        ns("generalplot_usubjid"),
+        "Select Subjects:",
+        choices = param_choices,
+        selected = param_choices[1],
+        multiple = TRUE,
+        options = list(`actions-box` = TRUE)
+      )
+    })
+
+    # select which variable to color the general line plot by
+    output$generalplot_colorby <- renderUI({
+      # deselect choices that are no pp parameters
+      param_choices <- sort(c("STUDYID", "PCSPEC", "ANALYTE", "DOSEA", "DOSNO", "USUBJID"))
+      pickerInput(
+        ns("generalplot_colorby"),
+        "Choose the variables to color by",
+        choices = param_choices,
+        selected = param_choices[length(param_choices)],
+        multiple = TRUE,
+        options = list(`actions-box` = TRUE)
+      )
+    })
+
+    # select the dose number/cycle for the general lineplot, if plotting by cycle is chosen
+    output$cycleselect <- renderUI({
+      req(input$generalplot_analyte)
+      y <- data() %>%
+        filter(ANALYTE == input$generalplot_analyte) %>%
+        pull(DOSNO) %>%
+        unique()
+      selectInput(ns("cycles"), "Choose the cycle :", choices = sort(y),
+                  multiple = TRUE, selected = y[1])
+
+    })
+
+    # render the general lineplot output in plotly
+    output$individualplot <- renderPlotly({
+      req(data())
+      req(input$generalplot_analyte)
+      req(input$generalplot_usubjid)
+      req(input$generalplot_colorby)
+      req(input$timescale)
+      req(input$log)
+
+      general_lineplot(
+        data(),
+        input$generalplot_analyte,
+        input$generalplot_usubjid,
+        input$generalplot_colorby,
+        input$timescale,
+        input$log,
+        cycle = input$cycles
+      )
+
+    })
+
+    # TAB: Mean Plot -----------------------------------------------------------
+
+    # This tabs plots the mean concentration of the input data in a dynamic plot
+
+    # select the analyte for the mean plot
+    output$analytemean <- renderUI({
+      y <- data() %>%
+        pull(ANALYTE) %>%
+        unique()
+      selectInput(ns("analytemean"), "Choose the Analyte:", choices = sort(y),
+                  selected = y[1])
+    })
+
+    # select the study id for the mean plot
+    # (needed for preclinical data, where multiple studyids per dataset)
+    output$studyidmean <- renderUI({
+      y <- data() %>%
+        pull(STUDYID) %>%
+        unique()
```

Comment:
suggestion: This pattern is repeated a lot in the app, might be worth putting it into a function in some kind of `utils` file? It is debatable if it is worth it, so whatever you feel more comfortable with.

```R
pull_unique <- function(data, colname) {
  data |> pull(colname) |> unique()
}
```

Reply:
Personally I think its not worth it for now, but if we end up using it even more later then we can implement it then